client/go/dogma/watch_service.go

```diff
@@ -0,0 +1,378 @@
+	select {
+	case latest := <-w.initialChan:
+		// Put it back to the channel so that this can return the value multiple times.
+		w.initialChan <- latest
+		return *latest
+	case <-time.After(timeout):
+		return Latest{Err: fmt.Errorf("failed to get the initial value. timeout: %v", timeout)}
+	}
+}
+
+// Latest returns the latest Revision and value of WatchFile() or WatchRepository() result.
+func (w *Watcher) Latest() Latest {
+	if latest := (*Latest)(atomic.LoadPointer(&w.latestPtr)); latest != nil {
+		return *latest
+	} else {
+		return Latest{Err: errors.New("latest is not set yet")}
+	}
+}
+
+// LatestValue returns the latest value of watchFile() or WatchRepository() result.
+func (w *Watcher) LatestValue() (interface{}, error) {
+	latest := w.Latest()
+	if latest.Err != nil {
+		return nil, latest.Err
+	}
+	return latest.Value, nil
+}
+
+// LatestValue returns the latest value of watchFile() or WatchRepository() result. If it's not available, this
+// returns the defaultValue.
+func (w *Watcher) LatestValueOr(defaultValue interface{}) interface{} {
+	latest := w.Latest()
+	if latest.Err != nil {
+		return defaultValue
+	}
+	return latest.Value
+}
+
+// Close stops watching the file specified in the Query or the pathPattern in the repository.
+func (w *Watcher) Close() {
+	atomic.StoreInt32(&w.state, stopped)
+	select {
+	case latest := <-w.initialChan:
+		w.initialChan <- latest
+	case <-time.After(100 * time.Millisecond):
+		w.initialChan <- &Latest{Err: errors.New("watcher is closed")}
+	}
+	w.scheduleCancelFunc() // After the first call, subsequent calls to a CancelFunc do nothing.
+	w.watchCancelFunc()
+}
+
+// Watch registers a func that will be invoked when the value of the watched entry becomes available or changes.
+func (w *Watcher) Watch(listener func(revision int, value interface{})) error {
+	if listener == nil {
+		return errors.New("listener is nil")
+	}
+	if w.isStopped() {
+		return errors.New("watcher is closed")
+	}
+
+	w.listenersMutex.Lock()
+	defer w.listenersMutex.Unlock()
+	w.updateListeners = append(w.updateListeners, listener)
+
+	if latest := w.Latest(); latest.Err == nil {
+		go func() {
+			// Perform initial notification so that the listener always gets the initial value.
+			listener(latest.Revision, latest.Value)
+		}()
+	}
+
+	return nil
+}
+
+func (ws *watchService) fileWatcher(projectName, repoName string, query *Query,
+	convertingValueFunc func(value interface{}) interface{}) (*Watcher, error) {
+	if query == nil {
+		return nil, errors.New("query should not be nil")
+	}
+
+	w := newWatcher(projectName, repoName, query.Path)
+	w.doWatchFunc = func(lastKnownRevision int) <-chan *WatchResult {
+		return ws.watchFile(w.watchCTX, projectName, repoName, strconv.Itoa(lastKnownRevision),
+			query, watchTimeout)
+	}
+	w.convertingResultFunc = func(result *WatchResult) *Latest {
+		var value interface{}
+		if convertingValueFunc != nil {
+			value = convertingValueFunc(result.Commit.Entries[0].Content)
```

Comment:
What should we do when something goes wrong with this user-provided function?

Reply:
I was thinking about removing this `convertingValueFunc` and passing the `WatchResult` as an argument so that a user can do whatever he or she wants in the `listener`.
Golang does not have generic and I have to use `interface{}` as the type to express the content and it's sort of annoying. WDYT?

Reply:
OK